README.md
tableone
========

[![R-CMD-check](https://github.com/kaz-yos/tableone/workflows/R-CMD-check/badge.svg)](https://github.com/kaz-yos/tableone/actions)
[![](https://www.r-pkg.org/badges/version/tableone)](https://www.r-pkg.org/pkg/tableone)
[![CRAN RStudio mirror
downloads](https://cranlogs.r-pkg.org/badges/tableone)](https://www.r-pkg.org/pkg/tableone)

**An R package to create “Table 1”, description of baseline
characteristics**

Creates “Table 1”, i.e., description of baseline patient
characteristics, which is essential in every medical research. Supports
both continuous and categorical variables, as well as p-values and
standardized mean differences. Weighted data are supported via the
survey package.

tableone was inspired by descriptive statistics functions in Deducer , a
Java-based GUI package by Ian Fellows. This package does not require GUI
or Java, and intended for command-line users.

tableone in action
==================

![screencast](man/figures/tableone.gif "screencast")

The code being executed can be found in the introduction vignette.

tableone code example
=====================

In this table, continuous and categorical variables can be placed in any
order. The p-valeus are from exact tests for pre-specified variables.
For nonnormal variables, it shows median and IQR instead of mean and SD,
and p-values are from nonparametric tests. Numerically coded categorical
variables can be transformed on the fly with factorVars. SMD stands for
standardized mean differences. For weighted data, first created a
svydesign object, and use the svyCreateTableOne() function. Most other
options remain the same.

    ## Load package
    library(tableone)
    ## Load data
    data(pbc, package = "survival")
    # drop ID from variable list
    vars <- names(pbc)[-1]
    ## Create Table 1 stratified by trt (can add more stratifying variables)
    tableOne <- CreateTableOne(vars = vars, strata = c("trt"), data = pbc,
                                factorVars = c("status","edema","stage"))
    ## Specifying nonnormal variables will show the variables appropriately,
    ## and show nonparametric test p-values. Specify variables in the exact
    ## argument to obtain the exact test p-values.
    print(tableOne, nonnormal = c("bili","chol","copper","alk.phos","trig"),
          exact = c("status","stage"), smd = TRUE,
          formatOptions = list(big.mark = ","))

    ##                          Stratified by trt
    ##                           1                           2                           p      test    SMD   
    ##   n                            158                         154                                         
    ##   time (mean (SD))        2,015.62 (1,094.12)         1,996.86 (1,155.93)          0.883          0.017
    ##   status (%)                                                                       0.884 exact    0.054
    ##      0                          83 (52.5)                   85 (55.2)                                  
    ##      1                          10 ( 6.3)                    9 ( 5.8)                                  
    ##      2                          65 (41.1)                   60 (39.0)                                  
    ##   trt (mean (SD))             1.00 (0.00)                 2.00 (0.00)             <0.001            Inf
    ##   age (mean (SD))            51.42 (11.01)               48.58 (9.96)              0.018          0.270
    ##   sex = f (%)                  137 (86.7)                  139 (90.3)              0.421          0.111
    ##   ascites (mean (SD))         0.09 (0.29)                 0.06 (0.25)              0.434          0.089
    ##   hepato (mean (SD))          0.46 (0.50)                 0.56 (0.50)              0.069          0.206
    ##   spiders (mean (SD))         0.28 (0.45)                 0.29 (0.46)              0.886          0.016
    ##   edema (%)                                                                        0.877          0.058
    ##      0                         132 (83.5)                  131 (85.1)                                  
    ##      0.5                        16 (10.1)                   13 ( 8.4)                                  
    ##      1                          10 ( 6.3)                   10 ( 6.5)                                  
    ##   bili (median [IQR])         1.40 [0.80, 3.20]           1.30 [0.72, 3.60]        0.842 nonnorm  0.171
    ##   chol (median [IQR])       315.50 [247.75, 417.00]     303.50 [254.25, 377.00]    0.544 nonnorm  0.038
    ##   albumin (mean (SD))         3.52 (0.44)                 3.52 (0.40)              0.874          0.018
    ##   copper (median [IQR])      73.00 [40.00, 121.00]       73.00 [43.00, 139.00]     0.717 nonnorm <0.001
    ##   alk.phos (median [IQR]) 1,214.50 [840.75, 2,028.00] 1,283.00 [922.50, 1,949.75]  0.812 nonnorm  0.037
    ##   ast (mean (SD))           120.21 (54.52)              124.97 (58.93)             0.460          0.084
    ##   trig (median [IQR])       106.00 [84.50, 146.00]      113.00 [84.50, 155.00]     0.370 nonnorm  0.017
    ##   platelet (mean (SD))      258.75 (100.32)             265.20 (90.73)             0.555          0.067
    ##   protime (mean (SD))        10.65 (0.85)                10.80 (1.14)              0.197          0.146
    ##   stage (%)                                                                        0.205 exact    0.246
    ##      1                          12 ( 7.6)                    4 ( 2.6)                                  
    ##      2                          35 (22.2)                   32 (20.8)                                  
    ##      3                          56 (35.4)                   64 (41.6)                                  
    ##      4                          55 (34.8)                   54 (35.1)

Installation
============

This version of tableone package for R is developmetal, and may not be
available from the CRAN. You can install it using one of the following
way.

**Direct installation from github**

You first need to install the devtools package to do the following. You
can choose from the latest stable version and the latest development
version.

    ## Install devtools (if you do not have it already)
    install.packages("devtools")
    ## Install directly from github (develop branch)
    devtools::install_github(repo = "kaz-yos/tableone", ref = "develop")

Using devtools may requires some preparation, please see the following
link for information.

<https://www.rstudio.com/projects/devtools/>

Contributors
============

I would like to thank all the contributors!

-   Alexander Bartel [ndevln](https://github.com/ndevln)
-   Jonathan J Chipman [chipmanj](https://github.com/chipmanj)
-   Justin Bohn [jmb01](https://github.com/jmb01)
-   Lucy D’Agostino McGowan
    [LucyMcGowan](https://github.com/LucyMcGowan)
-   Malcolm Barrett [malcolmbarrett](https://github.com/malcolmbarrett)
-   Rune Haubo B Christensen [runehaubo](https://github.com/runehaubo)
-   [gbouzill](https://github.com/gbouzill)

Similar or complementary projects
=================================

There are multiple similar or complementary projects of interest.

-   DescTools: Tools for Descriptive Statistics.
    <https://cran.r-project.org/web/packages/DescTools/index.html>
-   Gmisc: Descriptive Statistics, Transition Plots, and More.
    <https://cran.r-project.org/web/packages/Gmisc/>
-   Hmisc (summary.formula): Advanced table making and many more.
    <https://github.com/harrelfe/Hmisc/>
-   arsenal: An Arsenal of ‘R’ Functions for Large-Scale Statistical
    Summaries. <https://github.com/eheinzen/arsenal>
-   atable: Create Tables for Reporting Clinical Trials.
    <https://github.com/arminstroebel/atable>
-   compareGroups: Descriptive Analysis by Groups.
    <http://www.comparegroups.eu>
-   expss: Tables with Labels and Some Useful Functions from
    Spreadsheets and ‘SPSS’ Statistics.
    <https://github.com/gdemin/expss>
-   finalfit: Quickly Create Elegant Regression Results Tables and Plots
    when Modelling. <https://finalfit.org/index.html>
-   framework for easily create tables for reporting: framework for
    easily create tables for reporting.
    <https://davidgohel.github.io/flextable/>
-   furniture: Furniture for Quantitative Scientists.
    <https://cran.r-project.org/web/packages/furniture/>
-   gtsummary: Presentation-Ready Data Summary and Analytic Result
    Tables. <https://CRAN.R-project.org/package=gtsummary>
-   htmlTable: An R package for generating advanced tables.
    <https://github.com/gforge/htmlTable>
-   kableExtra: Construct Complex Table with ‘kable’ and Pipe Syntax.
    <https://github.com/haozhu233/kableExtra>
-   pander: An R Pandoc Writer. <https://rapporter.github.io/pander/>
-   pixiedust: Format models for console and to markdown, HTML, and
    LaTeX. <https://github.com/nutterb/pixiedust>
-   qwraps2: quickly placing data summaries and formatted regression
    results into .Rnw or .Rmd files.
    <https://github.com/dewittpe/qwraps2/>
-   stargazer: Well-Formatted Regression and Summary Statistics Tables.
    <https://cran.r-project.org/web/packages/stargazer/index.html>
-   tab: Functions for Creating Summary Tables for Statistical Reports.
    <https://cran.r-project.org/package=tab>
-   table1: Tables of Descriptive Statistics in HTML.
    <https://github.com/benjaminrich/table1>
-   table1xls: Exports Reproducible Summary Tables to Multi-Tab
    Spreadsheet Files.
    <https://cran.r-project.org/web/packages/table1xls/index.html>
-   xtable: Export Tables to LaTeX or HTML.
    <https://cran.r-project.org/web/packages/xtable/index.html>
-   (Python) tableone: Create “Table 1” for research papers in Python.
    <https://github.com/tompollard/tableone>


NEWS
tableone 0.13.2 (2022-04-15)
----------------------------------------------------------------

TEMPORARY BUG FIX

* Avoid calling && and || on a vector of length > 1 by disabling
  tests that call the 


tableone 0.13.1 (2022-03-27)
----------------------------------------------------------------

BUG FIX

* Fix some issues in the SMD vignette (@tszberkowitz)


tableone 0.13.0 (2021-07-04)
----------------------------------------------------------------

NEW FEATURES

* Capture the print argument in kableone (@ZWael)

BUG FIX

* Prepare for survey 4.1 using oldsvyquantile when available.
* Fix some vignette issues


tableone 0.12.0 (2020-07-25)
----------------------------------------------------------------

NEW FEATURE by by @ndevln (PR #69)

* Add formatOptions for the print method.
  formatOptions = list(big.mark = ",") to add marks like 1,000.


tableone 0.11.2 (2020-07-13)
----------------------------------------------------------------

BUG FIXES by @ndevln (PR #66)

* Hmisc labels no longer cause continuous vars to be treated as
  factors (Closes Issue #64)

* Setting value labels via labelled no longer result in tableone
  dropping variables.

ADDITIONAL BUG FIXES

* Add rmarkdown to Suggests.


tableone 0.11.1 (2020-03-07)
----------------------------------------------------------------

BUG FIXES

* Revise tests to avoid failures occuring only on ATLAS, MLK, and
  OpenBLAS.

* Drop methods from Imports as it is no longer needed.


tableone 0.11.0 (2020-03-01)
----------------------------------------------------------------

NEW FEATURES

* @ndevln contributed addOverall option, which add the overall
  column side-by-side with sratified columns when creating a
  table object.

BUG FIXES

* test-only dependency (suggests) on dummies was dropped as it
  produced warnings with model.matrix().

* In preparation for R 4.0.0, ModuleCreateTableForOneVar and
  svyCatSummaryForOneVar now use stringsAsFactors = FALSE.
  Some internal representation of data changed as a result.


tableone 0.10.0 (2019-02-16)
----------------------------------------------------------------

BUG FIXES

* Fix a missing value handling issue (via @chipmanj).
  Closes issue #26.

* mean (sd) was changed to mean (SD) to be consistent with
  median [IQR].

NEW FEATURES

* Add a helper for exporting TableOne objects to Markdown
  (via @malcolmbarrett). Currently, this is very basic. We hope
  to extend its functionalities.

* README.md now lists similar or complementary projects.


tableone 0.9.3 (2018-04-28)
----------------------------------------------------------------

BUG FIXES

* Fix incompatibility with lmerTest package (via @runehaubo).
* Suppressed "No sampling weights provided" warning on svrepdesign.


tableone 0.9.2 (2018-01-24)
----------------------------------------------------------------

BUG FIXES

* Address CRAN test issues on some platforms.
* Fix typos (via @LucyMcGowan)


tableone 0.9.1 (2017-11-06)
----------------------------------------------------------------

BUG FIXES

* ExtractSmd always returns a matrix even when there are only two
  groups for easier subsequent data manipulation.


tableone 0.9.0 (2017-11-03)
----------------------------------------------------------------

NEW FEATURES

* The "varLabels" option for the print.TableOne method was added.
  When TRUE, instead of printing the variable names, their
  corresponding variable labels are used. Variable labels must be
  stored in the data frame to be used via labelled::var_label
  function. This option is also available in ExtractSmd function.

* The "dropEqual" option for the print methods was implemented. If
  TRUE, the level description for two-level variables such as " = 1"
  and " = TRUE" are not shown. This can obscure what level is being
  shown depending on the variable naming scheme, thus, should only
  be used after the initial results were checked for correctness.

BUG FIXES

* Handling of lme4 models was improved in ShowRegTable.


tableone 0.8.2 (2017-08-31)
----------------------------------------------------------------

BUG FIXES

* Handle labelled class as categorical.


tableone 0.8.1 (2017-06-17)
----------------------------------------------------------------

BUG FIXES

* Fix alignment issue with the Missing column. Note currently the
  percentage is shown with 1 decimal and this is hard-coded.

* Change vignetteBuilder to knitr to form the index correctly.

* Single-quote package names in DESCRIPTION.


tableone 0.8.0 (2017-06-15)
----------------------------------------------------------------

NEW FEATURES

* The "missing" option for the print methods was implemented. If
  TRUE, a column called "Missing" is added as the rightmost column
  of the formatted table. This represents percentage of missing
  observation in each variable. Please note this is the percentage
  with respect to the unweighted raw observations even in weighted
  tables.

* The "padColnames" option was added the print.TableOne method. If
  TRUE, the column names of the formatted table become space-padded
  to center them.


tableone 0.7.6 (2016-07-12)
----------------------------------------------------------------

BUG FIXES

* The explanation for the "factorVars" argument for the functions
  CreateTableOne and svyCreateTableOne were changed for clarity.
  When factor variables are included in the argument, they are
  releveled to exlude empty levels. This was not clearly documented
  in the previous documentation. Thanks @eribul.

* svyrep.design objects (survey design objects with replicate weights)
  are allowed for the data argument in svyTableOne. This is considered
  experimental. Thanks @przemo.


tableone 0.7.5 (2016-04-10)
----------------------------------------------------------------

BUG FIXES

* ShowRegTable() now correctly supports models fit with geepack,
  nlme, and lme4.


tableone 0.7.4 (2016-03-31)
----------------------------------------------------------------

NEW FEATURE


R/modules-svy.R
################################################################################
### Modules for survey data support
##
## Created on: 2015-07-23
## Author: Kazuki Yoshida
################################################################################


###
### Generic helpers
################################################################################

## Function to return 1 as many as x elements
## Used for counting weighted sample size
one <- function(x) {
    rep(1, length(x))
}


## 0,1 indicator variable for missingness
miss <- function(x) {
    as.numeric(is.na(x))
}


## Return formula string having vars on RHS
FormulaString <- function(vars) {
    paste0(" ~ ", paste0(vars, collapse = " + "))
}


###
### Use oldsvyquantile when available
################################################################################

## 2020-07-30 via Thomas Lumley
## survey::svyquantile rewritten for survey 4.1
## To get the old behaviour you can simply call survey::oldsvyquantile() instead.
try_oldsvyquantile <- try(getFromNamespace(x = "oldsvyquantile", ns = "survey"))
if (is.function(try_oldsvyquantile)) {
    svyquantile <- survey::oldsvyquantile
}


###
### Helpers for both types of data
################################################################################

## Check for survey data; fail if not
StopIfNotSurveyDesign <- function(data) {

    if (!any(c("svyrep.design", "survey.design2", "survey.design") %in% class(data))) {
        stop("The data argument needs to be a survey design object.")
    }

}


###
### Helpers for continuous variable summary
################################################################################

svyN <- function(vars, design) {
    ## Same for all variables, just use first
    form <- sprintf(" ~ one(%s)", vars[1])
    ## No missing handling as all values are 1
    res <- svytotal(x = as.formula(form), design = design)
    ## Repeat for all variables
    out <- rep(res[1], length(vars))
    names(out) <- vars
    out
}


svyMiss <- function(vars, design) {

    ## Rewrite variables locally to missing 0,1 indicators
    design$variables[vars] <- lapply(design$variables[vars], miss)

    ## For each var find number of missingness
    res <- sapply(design$variables[vars], svytotal, design = design, na.rm=TRUE)
    out <- as.vector(res)
    names(out) <- vars
    out
}


svyPMiss <- function(vars, design) {
    svyMiss(vars, design) / svyN(vars, design)
}


svyMean <- function(vars, design) {

    ## For each var find weighted mean
    res <- sapply(design$variables[vars], svymean, design = design, na.rm=TRUE)
    out <- as.vector(res)
    names(out) <- vars
    out
}


svySd <- function(vars, design) {

    ## For each var find weighted variance (ie, not a variance, covariance matrix)
    res <- sapply(design$variables[vars],svyvar, design = design, na.rm=TRUE)
    out <- as.vector(sqrt(res))
    names(out) <- vars
    out
}


svyQuant <- function(vars, design, q = 0.5) {

    ## Use only one quantile
    ## Not as elegant as sapply, though sapply with svyquantile failed to run
    res <- vector()
    for ( i in 1:length(vars)) {
      var    <- vars[i]
      res[i] <- svyquantile(design$variables[var], design = design, quantiles = q[1], na.rm = TRUE)
    }
    out <- as.vector(res)
    names(out) <- vars
    out
}


svyContSummary <- function(vars, design) {

    ## Save for reuse
    nVec    <- svyN(vars, design)
    missVec <- svyMiss(vars, design)

    cbind(n      = nVec,
          miss   = missVec,
          p.miss = missVec / nVec * 100,
          mean   = svyMean(vars, design),
          sd     = svySd(vars, design),
          median = svyQuant(vars, design, q = 0.5),
          p25    = svyQuant(vars, design, q = 0.25),
          p75    = svyQuant(vars, design, q = 0.75),
          min    = svyQuant(vars, design, q = 0),
          max    = svyQuant(vars, design, q = 1))
}


###
### Helpers for categorical variable summary
################################################################################
## These work on one variable at a time

svyTable <- function(var, design) {
    form <- FormulaString(var)
    out <- svytable(formula = as.formula(form), design = design)
    out
}


svyLevel <- function(var, design) {
    names(svyTable(var, design))
}


svyPropTable <- function(var, design) {
    prop.table(svyTable(var, design))
}


## This one works at a single variable level within a stratum
svyCatSummaryForOneVar <- function(var, design) {

    ## Tables
    freqTab <- svyTable(var, design)
    propTab <- prop.table(freqTab)
    nLevels <- length(freqTab)
    ## Repeat as many as the levels
    nVec    <- rep(svyN(var, design), nLevels)
    missVec <- rep(svyMiss(var, design), nLevels)

    data.frame(n           = nVec,
               miss        = missVec,
               p.miss      = missVec / nVec * 100,
               ## Order must be as is.
               level       = names(freqTab),
               freq        = as.vector(freqTab),
               percent     = as.vector(propTab) * 100,
               cum.percent = cumsum(propTab) * 100,
               ## To protect against, level having <NA>
               row.names   = NULL,
               stringsAsFactors = FALSE)
}


## This one can take multiple variables and return a list
svyCatSummary <- function(vars, design) {

    sapply(vars, function(var) {

        svyCatSummaryForOneVar(var, design)

    }, simplify = FALSE)
}


###
### Helpers for testing for p values
################################################################################

## Function to do Wald test on a multi-degree variable after linear regression
svyGlmTermTest <- function(formula, design, test.terms, method = "Wald") {

    ## Perform linear regression and perform
    regTermTest(svyglm(formula, design), test.terms = test.terms, method = method)
}


## Given a formula string as its first argument, calls svyGlmTermTest correctly
svyTestNormal <- function(formulaString, design, test.terms, method = "Wald") {

    out <- svyGlmTermTest(formula = as.formula(formulaString), design = design,
                          test.terms = test.terms, method = method)
    ## Give an appropriate name for consistent extraction
    list(p.value = out$p[1,1])
}


## Kruskal-Wallis test
svyTestNonNormal <- function(formulaString, design) {

    ## This returns an htest object that has a scalar $p.value element
    svyranktest(formula = as.formula(formulaString), design = design)
}


svyTestChisq <- function(formulaString, design) {

    ## This returns an htest object that has a scalar $p.value element
    svychisq(formula = as.formula(formulaString), design = design)
}


svyModuleApproxExactTests <- function(data, vars, strataVarName,
                                      testApprox, argsApprox) {

    ## Loop over variables and create a list of xtabs
    ## Empty strata are kept in the corss tables. Different behavior than the cont counterpart!
    listXtabs <- sapply(X = vars,
                        FUN = function(var) {
                            ## Create a formula
                            formula <- as.formula(paste0("~ ", var, " + ", "..strataVar.."))

                            ## Create a 2-dimensional crosstable
                            svytable(formula = formula, design = data)
                        },
                        simplify = FALSE)

    ## Rename the second dimension of the xtabs with the newly create name.
    for (i in seq_along(listXtabs)) {

        names(dimnames(listXtabs[[i]]))[2] <- strataVarName
    }

    ## Loop over variables, and create p-values
    pValues <-
    sapply(X = vars,
           FUN = function(var) {
               formulaString <- paste0(" ~ ", var, " + ..strataVar..")
               ## Perform tests and return the result as 1x2 DF
               data.frame(pApprox = ModuleTestSafe(formulaString, testApprox,
                                                   c(list(design = data), argsApprox)),
                          pExact  = NA) # Not available for survey data. Just fill in with NA
           },
           simplify = FALSE)

    ## Create a single data frame (n x 2 (normal,nonormal))
    pValues <- do.call(rbind, pValues)

    ## Return both xtabs and p value df
    list(pValues = pValues, xtabs = listXtabs)
}
